### ISSUE TITLE
[Code Addition Request]: FlapAI: Flappy Bird Learns to Fly (with NEAT & DQN)

### ISSUE THREAD

**Issue by SK8-infi** [UTSAVS26/PyVerse#1621]
### Have you completed your first issue?

- [x] I have completed my first issue

### Guidelines

- [x] I have read the guidelines
- [x] I have the link to my latest merged PR

### Latest Merged PR Link

https://github.com/UTSAVS26/PyVerse/pull/1567

### 📌 Project Overview

**FlapAI** is an AI-powered version of **Flappy Bird**, where a bird **learns to survive longer** using **Neuroevolution (NEAT)** or **Deep Q-Learning**. The entire gameplay is simulated, and the AI gradually improves by **playing thousands of games** and evolving its strategy.

---

### 💡 Use Cases

* Hands-on intro to Neuroevolution and RL in games
* Showcases AI learning in dynamic environments
* Great demo project for portfolios and AI showcases

---

### 📂 Folder Structure

```
flapai/
│
├── game/
│   └── flappy_bird.py         # Pygame clone of Flappy Bird
│
├── agents/
│   ├── neat_agent.py          # NEAT algorithm using neat-python
│   ├── dqn_agent.py           # DQN implementation using PyTorch
│   └── base_agent.py
│
├── training/
│   ├── train_neat.py          # NEAT training loop
│   └── train_dqn.py           # DQN training loop
│
├── evaluation/
│   └── eval_visualize.py      # Watch trained agents play
│
├── assets/
│   ├── bird.png, pipe.png     # Game graphics
│
├── config/
│   └── neat-config.txt        # Config file for NEAT parameters
│
├── utils/
│   └── state_encoder.py       # Convert game state to agent input
│
├── README.md
├── requirements.txt
└── LICENSE
```

---

#### 🧬 NEAT (NeuroEvolution of Augmenting Topologies)

* No backprop — evolves neural nets over generations
* Fitness = survival time or score
* Uses `neat-python`

#### 🧠 DQN (Deep Q-Learning)

* Frame stacking as input
* Continuous state → discrete action (Flap or Not)
* Uses PyTorch with replay memory and target networks

---

### 🔧 Features

* 🕹️ Fully playable Flappy Bird clone using **Pygame**
* 🤖 NEAT & DQN support for training AI bots
* 📈 Live graph of training progress (fitness/score)
* 🎮 Visual mode: watch AI learn and improve in real time
* 📦 Easily swappable agents

---

### 📊 Sample Output

```
[NEAT] Gen 12 — Best Score: 35 — Avg Fitness: 22.6  
[DQN] Episode 200 — Avg Score (last 100): 27.2  
```

---

### 🧪 Enhancements

* Add genetic visualization tree (for NEAT)
* Add sound and game speed controls
* Train on 2D side-scroller clones or infinite runners
* Try curriculum learning (e.g., gradually decrease pipe gaps)

---

### 📘 README Highlights

**How it works:**

* The AI starts with no knowledge and gradually learns from **rewards** (e.g., +1 for passing a pipe, -100 for crashing).
* NEAT uses evolutionary strategies, while DQN uses gradient descent to learn an action-value policy.

**Dependencies:**

* Python, Pygame
* `neat-python` for NEAT
* PyTorch for DQN
* Matplotlib (for training visualization)

---

### Full Name

Shivansh Katiyar

### Participant Role

SSOC
